Somatic CNV tagging germline events, bringing AnnotatedInterval in line with tribble reading (and gearing up for writing as well), and region merging.

LeeTL1220 opened the pull request:
Most of these changes are to support automated evaluation of GATK CNV.

- Updates `AnnotatedIntervals` (formerly `SimpleAnnotatedGenomicRegion`) to use the tribble framework for reading.  Writing is done in a way that should be concordant with a future tribble writing framework, as per discussion with @droazen.
- Changes to `XsvLocatableTableCodec` to support usage of arbitrary config files.  This cannot be done when using tribble features in the CLI.  Already reviewed with @jonn-smith .  Support for SAM File headers and comments is included.
- *Note:*  The reading of `AnnotatedIntervals` cannot be done automatically on the command line, unless the config file is a sibling.  The tools below do not even attempt this, since the use cases involved will never have a sibling config file.
- Created a default config file in the jar file resources to read tsvs with locatable fields from the CNV collection files.  This is much less strict than the framework used by the CNV tools.  The reader will accept any columns (or subset of the columns).

CLIs (both experimental quality): 
- `TagGermlineEvents` is a simple tool that attempts to identify events in a tumor seg file that correspond to a germline events.  
  - This is done purely with concordance on the breakpoints of the events (within some padding).  
  - Input germline segments must have calls. 
  - If a germline call is broken into multiple segments, this tool will handle that appropriately (ditto if there are multiple tumor segments overlapping the germline call).

- `MergeAnnotatedRegions` will merge all overlapping regions and resolve annotation value conflicts.

Closes #3995

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegionCollection.java:
```diff
@@ -0,0 +1,261 @@
+     *
+     * @param input readable path to use for the xsv file.  Must be readable.  Never {@code null}.
+     * @param inputConfigFile config file for specifying the format of the xsv file.  Must be readable.  Never {@code null}.
+     * @param headersOfInterest Only preserve these headers.  These must be present in the input file.  This parameter should not include the locatable columns
```
Why `headersOfInterest` here and `columnsOfInterest` elsewhere?

LeeTL1220 replied:
Done

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegionCollection.java:
```diff
@@ -0,0 +1,261 @@
+     * @param headersOfInterest See {@link #create(Path, Path, Set)}
+     * @return See {@link #create(Path, Path, Set)}
+     */
+    public static SimpleAnnotatedGenomicRegionCollection create(final Path input, final Set<String> headersOfInterest) {
```
Another thought: some of the confusion here comes from the desired functionality of being able to read either `columnsOfInterest` or all columns.

Why not just always read all available columns in the file?  You'd only need a single relatively simple read method that knows how to dynamically parse the column header.

Then you could have a `subsetAnnotations` method in the record class that takes `columnsOfInterest` (although I probably wouldn't name the parameter as such) and returns a new record with the specified annotations (and throws an exception if an unavailable annotation is requested).  

You could have a corresponding method in the collections class, which would just be a one-liner with streams.

LeeTL1220 replied:
I have reduced the number of create methods.

I also made the create method that takes in a config file private.  This will enforce the input files having standardized contig, start, and end columns.

LeeTL1220 replied:
Done (with your approval)

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/CombineSegmentBreakpoints.java:
```diff
@@ -14,31 +17,29 @@
                 "This tool will load all segments into RAM.\n"+
-        "Expected interval columns are: " + SimpleAnnotatedGenomicRegion.CONTIG_HEADER + ", " +
-        SimpleAnnotatedGenomicRegion.START_HEADER + ", " + SimpleAnnotatedGenomicRegion.END_HEADER,
+                "Column headers for locatable information are taken from the first segment file.\n" +
```
You can specify the column headers or column index now, via config file.  By default these are `CONTIG`, `START`, and `END`.

Deleted comment.

Done.

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/CombineSegmentBreakpoints.java:
```diff
@@ -105,9 +107,20 @@ public void traverse() {
+                Arrays.asList(input1ToOutputHeaderMap, input2ToOutputHeaderMap), getBestAvailableSequenceDictionary(),
+                l -> progressMeter.update(l));
+
+        final SamFileHeaderMerger samFileHeaderMerger = new SamFileHeaderMerger(SAMFileHeader.SortOrder.coordinate,
```
Done.

LeeTL1220 replied:
Added docs and this tool will throw an exception if the dictionaries cannot be merged properly.  For example, identical contig names with different lengths will cause an exception.

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/xsvLocatableTable/XsvLocatableTableCodec.java:
```diff
@@ -137,9 +188,16 @@ else if ( split.size() > header.size() ) {
+                headerToIndex.get(finalEndColumn), header, split, dataSourceName);
     }
 
+    /**
```
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/xsvLocatableTable/XsvLocatableTableCodec.java:
```diff
@@ -256,9 +336,58 @@ private void readMetadataFromConfigFile(final Path configFilePath) {
+     */
+    public static void validateLocatableColumnName(String columnName) {
+        Utils.validateArg(!StringUtils.isEmpty(columnName), "column header is blank.");
+        Utils.validateArg(!org.apache.commons.lang.math.NumberUtils.isNumber(columnName), "column header cannot be a number: " + columnName);
```
Fixed and removed extraneous comments.

Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/CombineSegmentBreakpoints.java:
```diff
@@ -105,9 +107,20 @@ public void traverse() {
+                l -> progressMeter.update(l));
+
+        final SamFileHeaderMerger samFileHeaderMerger = new SamFileHeaderMerger(SAMFileHeader.SortOrder.coordinate,
+                Lists.newArrayList(simpleAnnotatedGenomicRegionCollection1.getSamFileHeader(),
```
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/xsvLocatableTable/XsvLocatableTableCodec.java:
```diff
@@ -149,22 +207,44 @@ else if ( split.size() > header.size() ) {
                         .collect(Collectors.toCollection(ArrayList::new));
+                headerToIndex = IntStream.range(0, header.size()).boxed()
+                        .collect(Collectors.toMap(i-> header.get(i), Function.identity()));
+                finalContigColumn = NumberUtils.isNumber(inputContigColumn) ? header.get(Integer.valueOf(inputContigColumn))
```
Typo from previous version of the code.  All using StringUtils
Done.

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/CombineSegmentBreakpoints.java:
```diff
@@ -105,9 +107,20 @@ public void traverse() {
+                Lists.newArrayList(simpleAnnotatedGenomicRegionCollection1.getSamFileHeader(),
+                        simpleAnnotatedGenomicRegionCollection2.getSamFileHeader()),true);
+
+        final List <String> finalAnnotations = Lists.newArrayList(finalList.get(0).getAnnotations().keySet());
```
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegion.java:
```diff
@@ -23,12 +14,9 @@
-    public final static String CONTIG_HEADER = "CONTIG";
-    public final static String START_HEADER = "START";
-    public final static String END_HEADER = "END";
+public final class SimpleAnnotatedGenomicRegion implements Locatable {
```
`AnnotatedInterval` it is.

LeeTL1220 replied:
Done

LeeTL1220 replied:
What about the class with the same name in the copy number package?

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegion.java:
```diff
@@ -59,6 +47,25 @@ public int getEnd() {
+     * @param annotationName the name for the annotation
+     * @param annotationValue the value
+     */
+    public void setAnnotation(final String annotationName, final String annotationValue) {
```
Deleted the setter.

LeeTL1220 replied:
`getAnnotations()` creates a copy.

LeeTL1220 replied:
Done.

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegion.java:
```diff
@@ -85,63 +92,9 @@ public String toString() {
-        } catch (final IOException ioe) {
-            throw new UserException.CouldNotCreateOutputFile("Cannot write file: " + outputFile.getAbsolutePath(), ioe);
-        }
+    public List<String> getAnnotationNames() {
```
Deleted.

Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegionCollection.java:
```diff
@@ -0,0 +1,261 @@
+ */
+public class SimpleAnnotatedGenomicRegionCollection {
+
+    static final Logger logger = LogManager.getLogger(SimpleAnnotatedGenomicRegionCollection.class);
```
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegionCollection.java:
```diff
@@ -0,0 +1,261 @@
+
+/**
+ * Represents a collection of annotated regions.  The annotations do not need to be known ahead of time, if reading from a file.
+ * Though {@link SimpleAnnotatedGenomicRegion::CONTIG_HEADER},
```
Done

LeeTL1220 commented on src/test/java/org/broadinstitute/hellbender/engine/MultiVariantWalkerIntegrationTest.java:
```diff
@@ -47,7 +47,7 @@ public void apply(
             count++;
 
-            //make sure we only move forward; if getSequenceDictionary returns null then there is only
+            //make sure we only move forward; if getSamFileHeader returns null then there is only
```
Not sure... Fixed.

Done.

LeeTL1220 commented on src/test/java/org/broadinstitute/hellbender/engine/ReferenceDataSourceUnitTest.java:
```diff
@@ -45,9 +44,9 @@ public void testNullReference() {
         try (ReferenceDataSource refDataSource = new ReferenceFileSource(TEST_REFERENCE)) {
             SAMSequenceDictionary sequenceDictionary = refDataSource.getSequenceDictionary();
-            Assert.assertEquals(sequenceDictionary.size(), 4, "Wrong number of sequences in sequence dictionary returned from refDataSource.getSequenceDictionary()");
+            Assert.assertEquals(sequenceDictionary.size(), 4, "Wrong number of sequences in sequence dictionary returned from refDataSource.getSamFileHeader()");
```
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegion.java:
```diff
@@ -59,6 +47,25 @@ public int getEnd() {
+        annotations.put(annotationName, annotationValue);
+    }
+
+
```
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegionCollection.java:
```diff
@@ -0,0 +1,261 @@
+     */
+    public void write(final File outputFile) {
+
+        Utils.validateArg(Files.isWritable(outputFile.toPath()), "Can not write to: " + outputFile.getAbsolutePath());
```
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegionWriter.java:
```diff
@@ -0,0 +1,91 @@
+            dataLine.set(endColumnHeader, record.getEnd());
+
+            // Now everything else.
+            record.getAnnotations().keySet().stream().forEach(k -> dataLine.set(k, record.getAnnotationValue(k)));
```
Could only get rid of `.stream()`.
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/AnnotatedGenomicRegionWriter.java:
```diff
@@ -0,0 +1,32 @@
+import java.io.Closeable;
+import java.util.List;
+
+public interface AnnotatedGenomicRegionWriter extends Closeable {
```
@droazen asked me to do this for future support of tribble writing.  @droazen , is this what you had in mind?

LeeTL1220 replied:
@droazen Wait a bit, I'm doing a refactor.

LeeTL1220 replied:
@droazen Was this file (`AnnotatedGenomicRegionWriter`) what you had in mind as a writer that would be easy to put into tribble?

LeeTL1220 replied:
Done (had conversation with @droazen offline)

LeeTL1220 commented on src/test/resources/org/broadinstitute/hellbender/tools/copynumber/utils/combine-segment-breakpoints-with-legacy-header.tsv:
```diff
@@ -1,4 +1,6 @@
-#SAMPLE_NAME=SAMPLE1
+@HD	VN:1.5
+@SQ	SN:1	LN:2000000
+@RG	ID:GATKCopyNumber	SM:sample1
```
I got rid of all `-with-legacy-header` ... 
Done

LeeTL1220 commented on src/test/resources/org/broadinstitute/hellbender/tools/copynumber/utils/combine-segment-breakpoints-with-legacy-header-learning-combined-copy-number.tsv:
```diff
@@ -1,3 +1,6 @@
+@HD	VN:1.5
+@SQ	SN:1	LN:2000000
+@RG	ID:GATKCopyNumber	SM:test-sample
```
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/xsvLocatableTable/XsvLocatableTableCodec.java:
```diff
@@ -256,9 +336,58 @@ private void readMetadataFromConfigFile(final Path configFilePath) {
         }
     }
 
+    /** TODO: Tests of this class and overridden config files.
```
Yes.  Done.

LeeTL1220 commented on src/test/resources/org/broadinstitute/hellbender/tools/copynumber/utils/combine-segment-breakpoints-different-annotation-headers-with-legacy-header.tsv:
```diff
@@ -1,4 +1,6 @@
-#SAMPLE_NAME=SAMPLE2
+@HD	VN:1.5
```
I have removed the `-with-legacy-headers` from the test file names.

Currently, since this is being driven by a config file (that contains CONTIG, START, and END), the input files are meant to represent files either generated by the CNV pipeline.  

These files could already conceivably have been generated by `CombineSegmentBreakpoints`.  

Are you asking me to test input files with different config file formats?

samuelklee replied:
My point was that your test files should primarily reflect actual inputs expected by the tool.

Now that you've changed over to config files that identify the column numbers of the locatable fields, I would not expect to see this file as a typical input---it's a mishmash of CNV 1.0/ReCapSeg-style columns, modified `CONTIG\tSTART\tEND` column headers, and a CNV-collections-style SAM header.

LeeTL1220 replied:
I was just using `Segment_Mean` and `Segment_Call` (CNV 1.0 style columns) as two example annotations.  While not likely an exact input (unless I was comparing CNV 1.0 to CNV ModeledSegments), the header can really be CONTIG, START, END + anything.  The only requirement for the columns for this tool is to have CONTIG, START, and END.

samuelklee replied:
I think that it is somewhat confusing to use this as test input if we'd never really see it in the real world.  As a developer who might have to maintain the tests that use this input, it's nice to be able to understand what the origin of the input might be (in case I have to modify it or come up with new inputs).  The name of the test file (*-different-annotation-headers-with-legacy-header.tsv) doesn't really provide enough clues in that regard, either.

Why not just use an unmodified ReCapSeg or CNV 1.0 seg file as test input?

LeeTL1220 replied:
Using fake data, I have included test files from JaBbA, UCSC, ReCapSeg (which is like CNV 1.0), PCAWG consensus files, and an Oncotator annotated CNV output.

Done.

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java:
```diff
@@ -1516,4 +1516,21 @@ static public int endOfShard(int shardIndex, int shardSize) {
+     * @param segment2  second segment to be joined
+     * @return          a new segment constructed by joining the two segments
+     */
+    public static SimpleInterval mergeSegments(final SimpleInterval segment1, final SimpleInterval segment2) {
```
Made it private.  It is still too specialized for general consumption in IntervalUtils.

Done.

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/CombineSegmentBreakpoints.java:
```diff
@@ -78,8 +77,11 @@
 
-        final List<SimpleAnnotatedGenomicRegion> segments1 = SimpleAnnotatedGenomicRegion.readAnnotatedRegions(segmentFiles.get(0), columnsOfInterest);
-        final List<SimpleAnnotatedGenomicRegion> segments2 = SimpleAnnotatedGenomicRegion.readAnnotatedRegions(segmentFiles.get(1), columnsOfInterest);
+        final SimpleAnnotatedGenomicRegionCollection simpleAnnotatedGenomicRegionCollection1 = SimpleAnnotatedGenomicRegionCollection.create(segmentFiles.get(0), columnsOfInterest);
```
Now it only requires a reference if neither input seg file contains a SAMFileHeader with a SAMSequenceDictionary.

LeeTL1220 replied:
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegionCollection.java:
```diff
@@ -0,0 +1,261 @@
+    static final Logger logger = LogManager.getLogger(SimpleAnnotatedGenomicRegionCollection.class);
+
+    public static final String ANNOTATED_REGION_DEFAULT_CONFIG_RESOURCE = "org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/annotated_region_default.config";
+    private SAMFileHeader samFileHeader;
```
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/AnnotatedGenomicRegionWriter.java:
```diff
@@ -0,0 +1,32 @@
+    /**
+     * Write only the header (and any SAMFileHeader or comments)
+     *
+     * @param SAMFileHeader SAM file header as a multiline string.  {@code null} is allowed, if not available.
```
Done.  Refactoring done here.

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegionWriter.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    @Override
+    public void writeHeader(final String SAMFileHeader, final List<String> comments, final List<String> annotations,
```
So, refactoring addresses first question.  

I'll add tests for second point.

LeeTL1220 replied:
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegionCollection.java:
```diff
@@ -0,0 +1,261 @@
+    /**
+     *  Same as {@link #create(Path, Path, Set)} , but uses the default annotation
+     *   region config file in the GATK.
+     * @param input See {@link #create(Path, Path, Set)}
```
Seems standard in other javadoc I see.  Should I leave it blank here?

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegionCollection.java:
```diff
@@ -0,0 +1,261 @@
+    }
+
+    /**
+     * Same as {@link #create(Path, Path, Set)} , but uses the default annotation
```
See above... made private
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegionCollection.java:
```diff
@@ -0,0 +1,261 @@
+    public static SimpleAnnotatedGenomicRegionCollection create(final List<SimpleAnnotatedGenomicRegion> regions,
+                                                                                    final SAMFileHeader samFileHeader,
+                                                                                    final List<String> annotations,
+                                                                                    final String contigColumnName,
```
Reduced to two public `create` methods.

Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegionCollection.java:
```diff
@@ -0,0 +1,261 @@
+                        codec.getComments(), regions, codec.getFinalContigColumn(), codec.getFinalStartColumn(), codec.getFinalEndColumn());
+
+            }
+            catch ( final FileNotFoundException ex ) {
```
It was a typo.  It meant that the input file was not found.

Done.

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedGenomicRegionCollection.java:
```diff
@@ -0,0 +1,261 @@
+    /** Does not include the locatable fields. */
+    private List<String> annotations;
+    private List<String> comments;
+    private List<SimpleAnnotatedGenomicRegion> records;
```
No longer mutable.

Effectively, there is enforcement at construction now.  And there is an attribute on the collection that will store the annotations that are in each AnnotatedInterval.

LeeTL1220 replied:
Done.

pshapiro4broad commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/AnnotatedInterval.java:
```diff
@@ -0,0 +1,88 @@
+ */
+public final class AnnotatedInterval implements Locatable  {
+
+    private SimpleInterval interval;
```
It looks like this can be `final`

LeeTL1220 replied:
@pshapiro4broad is here!  Now it's a party!  

Done.

pshapiro4broad commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedIntervalWriter.java:
```diff
@@ -0,0 +1,136 @@
+            // By initializing writer to be based on fileWriter, writer.close will close the fileWriter as well.
+            writer = new SimpleTableWriter(fileWriter, new TableColumnCollection(finalColumnList));
+        } catch (final IOException ioe) {
+            throw new UserException.CouldNotCreateOutputFile(outputFile, "Could not create: " + outputFile.getAbsolutePath());
```
Isn't this `new UserException.CouldNotCreateOutputFile(outputFile, ioe);`? It seems like the user would really want the exception in the output, as it could be a few different things.

LeeTL1220 replied:
Made it a GATKException.  Now user will get runtime error details.

Done.

pshapiro4broad commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/SimpleAnnotatedIntervalWriter.java:
```diff
@@ -0,0 +1,136 @@
+    // TODO: Test for optional SAM File Header
+    // TODO: Test for other column names
+    /**
+     * {@inheritDoc}
```
For an overridden method with no doc changes, it is sufficient to omit the javadoc completely, as javadoc automatically inherits docs for overridden methods.

LeeTL1220 replied:
That answers that.  Deleted.

Done.

pshapiro4broad commented on src/test/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/AnnotatedIntervalUtilsUnitTest.java:
```diff
@@ -0,0 +1,91 @@
+
+        return new Object[][] {
+            {
+                Arrays.asList(
```
You could avoid some code duplication here by moving the `Arrays.asList()` call into your test method.

LeeTL1220 replied:
Apologies if I am missing something, but it seems like I would have to do the following, which does not reduce code duplication:

```
    @DataProvider(name = "mergeTests")
    public Object [][] createMergeTests() {

        return new Object[][] {
            {
                    new AnnotatedInterval[]{
                            new AnnotatedInterval(new SimpleInterval("1", 100, 200),
                                    ImmutableSortedMap.of("Foo", "bar", "Foo1", "bar1")),
                            new AnnotatedInterval(new SimpleInterval("1", 100, 200),
                                    ImmutableSortedMap.of("Foo", "bar", "Foo1", "bar1"))
                    },
// ..........
```

No action

LeeTL1220 replied:
@pshapiro4broad Comment above...

pshapiro4broad replied:
You can avoid that by calling `new Object[][][] { ... }` E.g.
```
    @Test(dataProvider = "simpleTests")
    public void testSimpleTagging(AnnotatedInterval[] tumorSegments, AnnotatedInterval[] normalSegments,
                                  AnnotatedInterval[] gt) {

        final List<AnnotatedInterval> testResult = SimpleGermlineTagger.tagTumorSegmentsWithGermlineActivity(Arrays.asList(tumorSegments),
                Arrays.asList(normalSegments), "call",
                ReferenceUtils.loadFastaDictionary(new File(ReferenceUtils.getFastaDictionaryFileName(REF))), TEST_GERMLINE_TAGGING_ANNOTATION, 10);

        Assert.assertEquals(testResult, Arrays.asList(gt));
    }

    @DataProvider(name = "simpleTests")
    public Object[][] createSimpleTests() {
        return new Object[][][]{{
                {
                        // Tumor segments are assumed to be mutable.
                        new AnnotatedInterval(new SimpleInterval("1", 100, 200), Maps.newTreeMap(ImmutableSortedMap.of("call", "+"))),
                        new AnnotatedInterval(new SimpleInterval("1", 201, 300), Maps.newTreeMap(ImmutableSortedMap.of("call", "0")))
                },
                {
                        new AnnotatedInterval(new SimpleInterval("1", 100, 200), ImmutableSortedMap.of("call", "+")),
                        new AnnotatedInterval(new SimpleInterval("1", 201, 500), ImmutableSortedMap.of("call", "0"))
                },
                {
                        new AnnotatedInterval(new SimpleInterval("1", 100, 200), ImmutableSortedMap.of("call", "+", TEST_GERMLINE_TAGGING_ANNOTATION, "+")),
                        new AnnotatedInterval(new SimpleInterval("1", 201, 300), ImmutableSortedMap.of("call", "0", TEST_GERMLINE_TAGGING_ANNOTATION, "0"))
                }
        },
       ...
        };
    }
```

Although now that I've written it out it doesn't really format that much shorter. I'll leave it up to you.

LeeTL1220 replied:
Done

LeeTL1220 replied:
(no action)

pshapiro4broad commented on src/test/java/org/broadinstitute/hellbender/tools/copynumber/utils/germlinetagging/SimpleGermlineTaggerUnitTest.java:
```diff
@@ -0,0 +1,152 @@
+        return new Object[][] {
+                {
+                        // Trivial case
+                        Lists.newArrayList(
```
Why not `Arrays.asList()`? Do these need to be modifiable? Also, you could shorten the data provider by putting the array -> list conversion into the test method.

LeeTL1220 replied:
Converted to `Arrays.asList()`.  Done.

No action on the array -> list conversion in the test.

droazen commented:
@samuelklee: @LeeTL1220 and I just had a discussion about the writer aspect of this branch, and we agreed on the following:

1. Lee will introduce a new header type to encapsulate the information that's currently passed in individually to the `writeHeader()` method in `AnnotatedIntervalWriter`. This make the interface cleaner and more future-proof, since the signature will just become `writeHeader(AnnotatedIntervalHeader)`

2. Lee will start writing out 3 additional structured header lines (as comment lines) to every header, declaring the names of the chrom, start, and stop columns. These will not be respected on input yet (he will still be relying on a config file to get the names of these 3 columns), but it's the first step in the direction of storing all necessary schema information in the header of each file, rather than separately from each file.

3. Lee will file a github issue to eventually use these 3 header lines on input, when they are present, to get the names of the chrom/start/stop columns (possibly still with a fallback to a separate config file if they aren't, but that is a point we can debate in a future PR).

samuelklee commented:
Sorry @droazen @LeeTL1220, can you give me a bit more context?  @LeeTL1220 is no longer using any of the CNV-specific collections classes that I had hoped might be Tribble-ized in the future, so I'm OK with any decisions you guys make that is specific to his classes (does @jonn-smith have an opinion?)  I think that moving towards storing the config in the header is a good thing, in general.

If we need to make corresponding changes to the CNV-specific collections classes, then we should talk more.  Not all of those collections describe locatables, so I'm not sure how we could fit them in the Tribble framework.

LeeTL1220 commented:
@samuelklee I'm not making any changes to the CNV collection classes.  I
think none of this PR affects those classes.

On Thu, Mar 1, 2018 at 4:19 PM, samuelklee <notifications@github.com> wrote:

> Sorry @droazen <https://github.com/droazen> @LeeTL1220
> <https://github.com/leetl1220>, can you give me a bit more context?
> @LeeTL1220 <https://github.com/leetl1220> is no longer using any of the
> CNV-specific collections classes that I had hoped might be Tribble-ized in
> the future, so I'm OK with any decisions you guys make that is specific to
> his classes (does @jonn-smith <https://github.com/jonn-smith> have an
> opinion?) I think that moving towards storing the config in the header is a
> good thing, in general.
>
> If we need to make corresponding changes to the CNV-specific collections
> classes, then we should talk more. Not all of those collections describe
> locatables, so I'm not sure how we could fit them in the Tribble framework.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/broadinstitute/gatk/pull/4276#issuecomment-369734330>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACDXkzoWV1fcDEucTdcNZ_DggL0UW4M9ks5taGXhgaJpZM4Ru2it>
> .
>



-- 
Lee Lichtenstein
Broad Institute
75 Ames Street, Room 8011A
Cambridge, MA 02142
617 714 8632

LeeTL1220 commented:
@samuelklee Let's discuss tribble-ifying the CNV classes, but for this PR, leaving those alone.

LeeTL1220 commented:
@droazen number 3 in the list above:  #4489

LeeTL1220 commented:
@samuelklee There have been a lot of changes since I first opened this PR.  I thought you might want to check that I addressed the comments you had that are still applicable.  However, since this is now totally decoupled from the CNV Collection code, there should be less to look at.

@jonn-smith Can you look at the XsvLocatableCodec changes and the changes to the FuncotationFactory code?  And that I have honored what I said I would do in conversations.

samuelklee commented on src/test/resources/org/broadinstitute/hellbender/tools/copynumber/utils/combine-segment-breakpoints-comparison.tsv.seg:
```diff
@@ -0,0 +1,8 @@
+@HD	VN:1.5
+@SQ	SN:1	LN:2000000
+@RG	ID:GATKCopyNumber	SM:sample_gt
+CONTIG	START	END	Num_Probes	Segment_Mean	Segment_Call
```
This test file still seems unnatural to me (see original comment directly below).  How would this be generated in the real world?  It has a CNV-collections-style header and a mix of "CONTIG/START/END" and CNV 1.0 column headers, so it seems like it would have to be formatted manually.

Is the tool be able to consume truth files in their original format?  If so, then the test files should reflect that.

Also, note the filename extension is .tsv.seg---probably this can be just .seg.

samuelklee replied:
Same goes for most of the test files below.

What is the spec for these tools?  I think they should be able to consume **strictly unmodified** CNV-collections files (namely, the output of CallCopyRatioSegments) and truth files (CNV 1.0-style seg files) along with the appropriate config files.  Requiring that the user go in and add additional headers or modify column names is something we should avoid.

Perhaps this is what the tools already do, but if so, it's not clear to me from looking at these test files.

samuelklee commented on src/test/resources/org/broadinstitute/hellbender/tools/copynumber/utils/combine-segment-breakpoints-no-samheader.tsv.seg:
```diff
@@ -1,4 +1,5 @@
-#SAMPLE_NAME=SAMPLE1
+# This is another comment
+# This is yet another comment
```
This test file is also not up to date.  The CNV-collections column headers are correct, but the SAM-style header is missing.  (I think you probably generated this file when the header consisted only of `#SAMPLE_NAME=...` and didn't update it when the CNV-collections format was updated?)

LeeTL1220 replied:
As per discussions with @jonn-smith and @droazen , the AnnotatedInterval codec will support files that either contain comments (`#`) or a SAM File Header ('@').  Not both.  Then the locatable column headers is whatever is specified by the user config file or the default config file.  The default config file will match what the CNV collection classes use.

This test is of a file that is comments only and uses the default configuration file.  Therefore, the test should not have a SAM File Header and the locatable column headers should be CONTIG, START, and END.

samuelklee replied:
When would this file appear naturally in the wild?  My point is that it wouldn't (it would require that a user take a CNV-collections file, strip the SAM header, and add comments), so it's extremely confusing to base a test on it.  Test files should reflect the most typical use cases; if a file is atypical or unusual (e.g., an improperly or unexpectedly formatted file meant to stress test the code), then it should be clearly named as such.

Can you list here exhaustively the typical formats expected for each of the tools?  Ideally, I would be able to easily discern this just from glancing at the test files, but my main objection is that I cannot---precisely because most of them are a mishmash of conventions.

samuelklee replied:
Also, I think it might be more convenient to make the default config file specify the locatable column headers by indices 0-2.  For seg files that have `Sample` as the first column, I think it's a bit cleaner to just use `cut` externally than it is to change column headers with e.g. `sed`.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/annotatedregion/AnnotatedInterval.java:
```diff
@@ -0,0 +1,102 @@
+/**
+ * Simple class that just has an interval and sorted name-value pairs.
+ */
+public final class AnnotatedInterval implements Locatable, Feature {
```
The package name should be changed to reflect the name change to `AnnotatedInterval`.  You should probably also edit the commit message for this branch.

Also note https://github.com/broadinstitute/gatk/issues/3884.  I don't mind the redundancy for now.

LeeTL1220 replied:
Renamed the package.

Done.

samuelklee commented:
@LeeTL1220 Apologies, I don't think I have the bandwidth for a detailed re-review, but I'm OK with the XSV code if @jonn-smith approves.

However, I still don't understand the test files and left some comments there.

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/xsv/LocatableXsvFuncotationFactory.java:
```diff
@@ -187,70 +187,41 @@ public void setSupportedFuncotationFields(final List<Path> inputDataFilePaths) {
+                            throw new UserException.MalformedFile("Could not decode from data file: " + dataPath.toUri().toString());
+                        }
+
+                        supportedFieldNames.addAll(codec.getHeaderWithoutLocationColumns());
```
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/xsv/LocatableXsvFuncotationFactory.java:
```diff
@@ -187,70 +187,41 @@ public void setSupportedFuncotationFields(final List<Path> inputDataFilePaths) {
 
                     // Adjust the manual annotations to make sure we don't try to annotate any fields we aren't
                     // responsible for:
+                    //TODO: Isn't this the default map not the override map as the name would imply?
```
Deleting the TODO then...
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/xsvLocatableTable/XsvLocatableTableCodec.java:
```diff
@@ -78,29 +112,79 @@
     /** The XSV Table Header */
     private List<String> header;
 
+    /** The XSV Table Header */
```
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/xsvLocatableTable/XsvLocatableTableCodec.java:
```diff
@@ -135,35 +219,88 @@ else if ( split.size() > header.size() ) {
         }
-
-        return new XsvTableFeature(contigColumn, startColumn, endColumn, header, split, dataSourceName);
+        // HACK: For tribble to work properly, we need to know the position (in bytes) of the file pointer.  Currently, that cannot be ascertained.
```
I've removed the null check.

Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/xsvLocatableTable/XsvLocatableTableCodec.java:
```diff
@@ -135,35 +219,88 @@ else if ( split.size() > header.size() ) {
     }
 
+    /**
+     * Read until we get to the header of this xsv
```
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/xsvLocatableTable/XsvLocatableTableCodec.java:
```diff
@@ -135,35 +219,88 @@ else if ( split.size() > header.size() ) {
+                finalEndColumn = StringUtils.isNumeric(inputEndColumn) ? header.get(Integer.valueOf(inputEndColumn))
+                        : determinePrefixForHeader() + inputEndColumn;
 
+                locatableColumns = Lists.newArrayList(finalContigColumn, finalStartColumn, finalEndColumn);
```
Done

LeeTL1220 replied:
(At one time I was modifying these and I forgot to adjust this line when that was no longer true)

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/xsvLocatableTable/XsvLocatableTableCodec.java:
```diff
@@ -241,23 +382,150 @@ private void readMetadataFromConfigFile(final Path configFilePath) {
+     * Throw an exception if the given column name cannot be used for one of the locatable columns.
+     * @param columnName candidate column name for one of the locatable fields (contig, start, or end)
+     */
+    public static void validateLocatableColumnName(String columnName) {
```
Done

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/xsvLocatableTable/XsvLocatableTableCodec.java:
```diff
@@ -241,23 +382,150 @@ private void readMetadataFromConfigFile(final Path configFilePath) {
+     */
+    public SAMFileHeader renderSamFileHeader() {
+        assertHeaderInitialized();
+        if (isPreambleSamFileHeader(getPreamble())) {
```
Done

LeeTL1220 replied:
I would hope that the optimizer would catch this, but it is explicit now.

LeeTL1220 commented on src/main/java/org/broadinstitute/hellbender/utils/codecs/xsvLocatableTable/XsvLocatableTableCodec.java:
```diff
@@ -241,23 +382,150 @@ private void readMetadataFromConfigFile(final Path configFilePath) {
+        final SAMFileHeader result = codec.decode(reader, null);
+
+        final List<String> finalComments = new ArrayList<>();
+        finalComments.addAll(result.getComments());
```
I have added a comment to that effect.

Done

LeeTL1220 commented:
Victory!!
